Migrate thunk package to ESM

markerikson opened the pull request:
This PR:

- Migrates the `redux-thunk` package to be full ESM with CJS compat

Also see https://github.com/reduxjs/redux-toolkit/pull/3095

mgol commented on package.json:
```diff
@@ -30,7 +39,7 @@
     "format": "prettier --write {src,test,typescript_test}/**/*.{js,ts}",
     "format:check": "prettier --check {src,test,typescript_test}/**/*.{js,ts}",
-    "lint": "eslint '{src,test,typescript_test}/**/*.{js,ts}'",
+    "lint": "eslint {src,test,typescript_test}/**/*.{js,ts}",
```
This reversed my change from https://github.com/reduxjs/redux-thunk/pull/339. Any reason for that?

markerikson replied:
@mgol : hmm. It doesn't seem to work on Git Bash on Windows, which is what _I_ normally use :)

Wonder if escaping double quotes would work better here?

mgol replied:
Ah, the choose your poison dilemma… 😅 

Yes, double quotes work on macOS as well! I’ve just checked.

markerikson commented:
Wrote this over in Discord, but let me paste here to explain what's going on:

> it seems, based on both experimentation and anecdotal discussion, that Jest does not like it when mixed ESM+CJS packages have a default export
> 
> I can do some more checking and try to narrow down the exact conditions
> 
> but that's why our Jest tests were failing earlier on the ESM migration branch
> 
> once I made RTK's package be `{type: "module"}`, `redux-thunk` started getting imported into Jest as `{default: thunk}`, rather than just thunk
> 
> ie, an object with the middleware function inside, not the middleware function itself
> 
> and that caused every use of configureStore to explode with "hey one of these things in this middleware array is not a function"
I changed `redux-thunk` to have actual exports in package.json, but left the default export as-is.  published that as an actual `3.0.0-alpha.0`.  that still errored.
> 
> changed it to be `export thunk, export withExtraArgument`, removed the default export, published that as `3.0.0-alpha.1`.
> that fixed things in RTK's tests
> and then Immer started failing with the same problem
> 
> given that the migration we're talking about is going from, roughly, `import thunk` to `import { thunk }`, I'm not concerned about that being a pain. doubly so because they should be using RTK and not doing it themselves

So, my intent for now _is_ to remove the default export from 3.0.

I do want to do more testing, but this seems like a probably-necessary change for the best tooling compat.

markerikson commented:
As a related addendum: Michel Weststrate is looking at dropping Immer's default export for Immer 10.0:

- https://github.com/immerjs/immer/issues/1015
